subprojects/core/src/main/java/org/gradle/api/internal/initialization/transform/CollectDirectClassSuperTypesTransform.java

```diff
@@ -0,0 +1,124 @@
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.initialization.transform;
+
+import com.google.common.collect.Ordering;
+import org.gradle.api.artifacts.transform.InputArtifact;
+import org.gradle.api.artifacts.transform.TransformAction;
+import org.gradle.api.artifacts.transform.TransformOutputs;
+import org.gradle.api.artifacts.transform.TransformParameters;
+import org.gradle.api.file.FileSystemLocation;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.internal.classpath.ClasspathWalker;
+import org.gradle.internal.file.FileException;
+import org.gradle.internal.file.Stat;
+import org.gradle.work.DisableCachingByDefault;
+import org.objectweb.asm.ClassReader;
+
+import javax.inject.Inject;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static com.google.common.collect.ImmutableSortedSet.toImmutableSortedSet;
+
+/**
+ * TODO: This class has similar implementation in build-logic/packaging/src/main/kotlin/gradlebuild/instrumentation/transforms/CollectDirectClassSuperTypesTransform.kt.
+ *  We could reuse the same class at some point.
+ */
+@DisableCachingByDefault(because = "Not worth caching.")
+public abstract class CollectDirectClassSuperTypesTransform implements TransformAction<TransformParameters.None> {
+
+    private static final Predicate<String> ACCEPTED_TYPES = type -> type != null && !type.startsWith("java/lang");
+    public static final String SUPER_TYPES_MARKER_FILE_NAME = ".gradle-super-types.marker";
+    private static final String FILE_SUFFIX = ".super-types";
+
+    @Inject
+    public abstract ObjectFactory getObjects();
+
+    @PathSensitive(PathSensitivity.NAME_ONLY)
+    @InputArtifact
+    public abstract Provider<FileSystemLocation> getInput();
+
+    public void transform(TransformOutputs outputs) {
+        try {
+            InjectedInstrumentationServices services = getObjects().newInstance(InjectedInstrumentationServices.class);
+            ClasspathWalker walker = services.getClasspathWalker();
+            File inputFile = getInput().get().getAsFile();
+            Map<String, Set<String>> superTypes = new TreeMap<>();
+            try {
+                walker.visit(inputFile, entry -> {
+                    if (entry.getName().endsWith(".class")) {
+                        ClassReader reader = new ClassReader(entry.getContent());
+                        String className = reader.getClassName();
+                        Set<String> classSuperTypes = getSuperTypes(reader);
+                        if (!classSuperTypes.isEmpty()) {
```

Comment:
Here we read super types for every class and we don't build type hierarchy yet.

So if we have:

```
class A extends B {}

and

class C {}
```

then we would add here to a map `"A" -> {"B"}`, e.g.:
``` 
superTypes.put("A", {"B"});
```

but we won't add `"C" -> {}`, since `C` supertypes are an empty set.

Reply:
Ah, sorry, I was misreading the code. Makes sense. All good!